src/runtime/ndarray.cc

```diff
@@ -206,6 +211,46 @@ void NDArray::CopyFromTo(DGLArray* from, DGLArray* to) {
       from->dtype);
 }
 
+void NDArray::RecordedCopyFromTo(DGLArray* from, DGLArray* to, void* pyt_ctx) {
+  size_t from_size = GetDataSize(*from);
+  size_t to_size = GetDataSize(*to);
+  CHECK_EQ(from_size, to_size)
+      << "DGLArrayCopyFromTo: The size must exactly match";
+
+  CHECK(
+      from->ctx.device_type == to->ctx.device_type ||
+      from->ctx.device_type == kDGLCPU || to->ctx.device_type == kDGLCPU)
+      << "Can not copy across different ctx types directly";
+
+  // Use the context that is *not* a cpu context to get the correct device
+  // api manager.
+  DGLContext ctx = from->ctx.device_type != kDGLCPU ? from->ctx : to->ctx;
+  CHECK(ctx.device_type == kDGLCUDA)
+      << "Can not record event if not cuda device";
+  DeviceAPI::Get(kDGLCUDA)->RecordedCopyDataFromTo(
+      from->data, static_cast<size_t>(from->byte_offset), to->data,
+      static_cast<size_t>(to->byte_offset), from_size, from->ctx, to->ctx,
+      from->dtype, pyt_ctx);
+}
+
+NDArray NDArray::PinnedEmpty(
+    std::vector<int64_t> shape, DGLDataType dtype, DGLContext ctx) {
+  CHECK_EQ(ctx.device_type, kDGLCPU) << "Only NDArray on CPU can be pinned";
+  NDArray ret = Internal::Create(shape, dtype, ctx);
+  // setup memory content
+  size_t size = GetDataSize(ret.data_->dl_tensor);
+  // size_t alignment = GetDataAlignment(ret.data_->dl_tensor);
+  if (size > 0) {
+    ret.data_->dl_tensor.data = DeviceAPI::Get(kDGLCUDA)->AllocPinnedDataSpace(
+        size, &(ret.data_->pyt_ctx), &(ret.data_->pyt_raw_deleter));
+    CHECK(
+        ret.data_->pyt_ctx != nullptr && ret.data_->pyt_raw_deleter != nullptr)
+        << "Can not return proper CachingHostAllocator";
```

Comment:
Maybe just "Data allocation failed."? From the code, I can't tell the failed is caused by CachingHostAllocator or other issue.

Reply:
Updated comment. Actually, the previous call `AllocPinnedDataSpace` changes the two pointers `pyt_ctx` and `pyt_raw_deleter`; and this is to check the return status of `AllocPinnedDataSpace` (CUDACachingAllocator).